src/olm/account/mod.rs

```diff
@@ -682,16 +790,171 @@ mod libolm {
+                key_id: KeyId(0),
+                key: Curve25519SecretKey::from_slice(&key.private_key),
+                published: true,
+            }
+        }
+    }
+
+    impl TryFrom<&FallbackKey> for OneTimeKey {
+        type Error = ();
+
+        fn try_from(key: &FallbackKey) -> Result<Self, ()> {
+            Ok(OneTimeKey { private_key: key.secret_key().to_bytes() })
+        }
+    }
+
+    #[derive(Zeroize, ZeroizeOnDrop)]
+    pub(crate) struct OptFallbackKey {
+        pub(crate) fallback_key: Option<OneTimeKey>,
+    }
+
+    impl Decode for OptFallbackKey {
+        fn decode(reader: &mut impl std::io::Read) -> Result<Self, DecodeError> {
+            let present = bool::decode(reader)?;
+
+            let fallback_key = if present {
+                let fallback_key = OneTimeKey::decode(reader)?;
+
+                Some(fallback_key)
+            } else {
+                None
+            };
+
+            Ok(Self { fallback_key })
+        }
+    }
+
+    impl Encode for OptFallbackKey {
+        fn encode(&self, writer: &mut impl std::io::Write) -> Result<usize, EncodeError> {
+            let ret = match &self.fallback_key {
+                None => false.encode(writer)?,
+                Some(key) => {
+                    let mut ret = true.encode(writer)?;
+                    ret += key.encode(writer)?;
+
+                    ret
+                }
+            };
+
+            Ok(ret)
+        }
+    }
+
+    #[derive(Encode, Decode, Zeroize, ZeroizeOnDrop)]
+    /// Pickle used for dehydrated devices.
+    ///
+    /// Dehydrated devices are used for receiving encrypted messages when the
+    /// user has no other devices logged in, and are defined in
+    /// [MSC3814](https://github.com/matrix-org/matrix-spec-proposals/pull/3814).
+    pub(super) struct Pickle {
+        version: u32,
+        #[secret]
+        private_curve25519_key: Box<[u8; 32]>,
+        #[secret]
+        private_ed25519_key: Box<[u8; 32]>,
+        one_time_keys: Vec<OneTimeKey>,
+        opt_fallback_key: OptFallbackKey,
+    }
+
+    pub(super) const PICKLE_VERSION: u32 = 1;
+
+    impl From<&Account> for Pickle {
+        fn from(account: &Account) -> Self {
+            let one_time_keys: Vec<_> = account
+                .one_time_keys
+                .secret_keys()
+                .iter()
+                .map(|(_key_id, secret_key)| OneTimeKey { private_key: secret_key.to_bytes() })
+                .collect();
+
+            let fallback_key =
+                account.fallback_keys.fallback_key.as_ref().and_then(|f| f.try_into().ok());
+
+            Self {
+                version: PICKLE_VERSION,
+                private_curve25519_key: account.diffie_hellman_key.secret_key().to_bytes(),
+                private_ed25519_key: account
+                    .signing_key
+                    .unexpanded_secret_key()
+                    .expect("Cannot dehydrate an account created from a libolm pickle"),
```

Comment:
Actually, @uhoreg, won't this pose a problem? In any case I don't think we should use `expect()` here.

Reply:
So the expectation is that we ourselves will always create a new Account when we want to create a dehydrated device, but people may misuse this.

I think we should convert this into a `TryFrom` and let the person be aware of this limitation. In the Rust SDK itself we can guarantee that we don't misuse the API and call `expect()` there.